### PR TITLE
[GOBBLIN-2167] Allow filtering of Hive datasets by underlying HDFS folder location

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2167

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
Hive tables can be located in different folders in HDFS even if they belong to the same database.
This becomes tricky to manage within a single Gobblin job especially when there are different permissions and handling based on underlying files for viewFS.

This PR adds a configuration to have a regex to filter tables based on their table location:
```
hive.dataset.tableFolderAllowlistFilter=<regex>
```
where tables with paths matching this filter will be selected, otherwise ignore

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Unit tests

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinderTest.java [hunk 1]
```diff
@@ -215,6 +215,32 @@ public void testDatasetConfig() throws Exception {
...
+    // Try a regex with multiple groups
+    properties.put(HiveDatasetFinder.TABLE_FOLDER_ALLOWLIST_FILTER, "(/tmp/|a).*");
+
+    HiveDatasetFinder finder = new TestHiveDatasetFinder(FileSystem.getLocal(new Configuration()), properties, pool);
+    List<HiveDataset> datasets = Lists.newArrayList(finder.getDatasetsIterator());
+
+    Assert.assertEquals(datasets.size(), 1);
+
+    properties.put(HiveDatasetFinder.HIVE_DATASET_PREFIX + "." + WhitelistBlacklist.WHITELIST, "");
+    // The table located at /tmp/test should be filtered
```
Comment (khandelwal-prateek) [inline gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinderTest.java:235]: it would be helpful to call out that the dataset table is created at path `/tmp/test` on line 221 since there is another assertion above using a different regex which doesn't filter the table

File: gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinderTest.java [hunk 2]
```diff
@@ -215,6 +215,32 @@ public void testDatasetConfig() throws Exception {
...
+    properties.put(HiveDatasetFinder.TABLE_FOLDER_ALLOWLIST_FILTER, "/a/b");
+
+    finder = new TestHiveDatasetFinder(FileSystem.getLocal(new Configuration()), properties, pool);
+    datasets = Lists.newArrayList(finder.getDatasetsIterator());
+
+    Assert.assertEquals(datasets.size(), 0);
```
Comment (khandelwal-prateek) [inline gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinderTest.java:241]: can also add a test for the case where the regex is empty or null

File: gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinderTest.java [hunk 3]
```diff
@@ -215,6 +215,32 @@ public void testDatasetConfig() throws Exception {
 
   }
 
+  @Test
+  public void testHiveTableFolderFilter() throws Exception {
```
Comment (khandelwal-prateek) [inline gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinderTest.java:219]: can be renamed to `testHiveTableFolderAllowlistFilter`

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinder.java [hunk 4]
```diff
@@ -294,6 +304,12 @@ protected HiveDataset computeNext() {
     };
   }
 
+  protected static boolean shouldAllowTableLocation(Optional<String> regex, Table table) {
+    if (!regex.isPresent()) {
+      return true;
+    }
+    return Pattern.compile(regex.get()).matcher(table.getSd().getLocation()).matches();
```
Comment (khandelwal-prateek) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinder.java:311]: Pattern.compile(regex.get()) is called every time the method shouldAllowTableLocation is executed when the regex is present.  Compiling a regex every time can be inefficient, we can compile the pattern once and store it, eg.:
`private static Pattern compiledAllowlistPattern = regex.map(Pattern::compile).orElse(null);`

Comment (Will-Lo) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveDatasetFinder.java]: I compiled it as part of the class field instead, good callout

**Comment by khandelwal-prateek** [comment]
LGTM!
